Adding information about Templating in Fleet. (#256)
rancher/fleet-product-docs: community-docs/next/modules/ROOT/pages/how-tos-for-users/template.adoc

= {product_name} Templating
:revdate: 2026-05-14
:page-revdate: {revdate}

> :page-aliases: /template

Reviewer: Wouldn't it make more sense for the alias to be `/templating`?

Author: Fxied.

Templating helps you customize deployments for different clusters without maintaining separate configuration files. This helps you:

* Deploy environment-specific image tags.
* Use cluster labels to define regional configurations.
* Generate dynamic values for Helm charts.
* Reuse computed values across multiple manifests.

{product_name} preprocesses the following files:

> * ++fleet.yaml++

Reviewer: This does not render with a monospace font.
```suggestion
* `fleet.yaml`
```
Same remark about other words enclosed with `++` throughout this patch.

Author: Fixed.

* Helm values files
* Kubernetes manifests
* HelmOps resources

> When you install Helm charts using {product_name}. It is important to distinguish between {product_name}-managed files and internal Helm chart files:

Reviewer 2: it should be `{product_name}, it is important..`, right?

* *Helm chart ++values.yaml++ (No templating):*
** The default ++values.yaml++ file bundled inside a Helm chart does not support {product_name} templating. Any substitutions happen during the standard Helm chart rendering phase, before {product_name} applies the chart. If you use ++${var}++ in this file, Helm treats it as plain text.
* *Fleet ++valuesFiles++ and ++values++ (Templating supported):*
** {product_name} allows you to reference additional values files through ++fleet.yaml++ using the ++valuesFiles++ or ++values++ keys. These {product_name}-specific configurations fully support templating, allowing you to apply dynamic, environment-specific overrides.

== How {product_name} templating works

Before a ++Bundle++ is created, the Fleet controller scans specific files (like ++fleet.yaml++ and ++values.yaml++). It searches for ++${ }++ delimiters. These are used instead of standard Go-template ++{{ }}++ to avoid collisions with Helm's own templating engine.

[source,mermaid]

> ====

Reviewer: This graph is not rendered, at least not on my local build.

graph LR
    A[Git Repository] -->|"fleet.yaml / manifests"| B(Fleet Controller)
    B -->|"Stage 1: Pre-processing"| C{Template Engine}
    C -->|"Evaluates ${ }"| D[Processed BundleDeployment]
    D -->|"Distributed to"| E(Downstream Fleet Agent)
    E -->|"Stage 2: Helm Execution"| F{Helm Engine}
    F -->|"Evaluates {{ }}"| G[Live K8s Resources]
====

For example, if you have a ++values.yaml++ file with:

[source,yaml]
====
values:
  region: ${ .ClusterLabels.region }

> ====

Reviewer: This does not render as a code block on my machine.
The following, however, does:
```suggestion
[,yaml]
----
values:
  region: ${ .ClusterLabels.region }
----
```

Same remark about other code blocks throughout this patch.

Author: All code blocks fixed

> If the target cluster contains the label: ++region: us-west++. Then the {product_name} generates:

Reviewer: ```suggestion
If the target cluster contains the label: ++region: us-west++. Then {product_name} generates:
```

| Rendering (Cluster side)

| *Context*
| Cluster Labels, Annotations, Name
| Values.yaml, Chart metadata

| *Processor*
| Fleet Controller
| Helm Binary

|===

For example, if you have a ++fleet.yaml++ with:

[source,yaml]
====
values:
  cluster: ${ .ClusterName }

config:
  message: "{{ .Values.message }}"
====

In this example:

* {product_name} resolves ++${ .ClusterName }++
* Helm later resolves ++{{ .Values.message }}++

== ++Value precedence++

When {product_name} prepares the final values for a deployment, it merges configuration from multiple sources. If the same key exists in multiple places, the source with the highest precedence wins.

. `templateValues`: Values generated here always overwrite other sources. Use this for logic-heavy variables.
. `targetCustomizations`: Overrides defined for specific cluster groups.
. `helm.values`: Global values defined in your {product_name}.yaml.
. `helm.valuesFrom`: Values loaded dynamically from ConfigMaps or Secrets on downstream clusters.
. `helm.valuesFiles`: External YAML files referenced in the helm block.
. `Chart values.yaml`: The default values bundled inside the Helm chart.

[source, mermaid]
====

> graph LR

Reviewer 2: Seems like you add the helm.valuesFrom above but the mermaid was neglected. is that intentional?

Reviewer: Also, the graph does not render.

[source,yaml]
====
helm:
  templateValues:
    # 1. Define once
    templateValues:
      environment: production
      region: eu-west
      enableMetrics: "true"
  values:
    values:
      app:
        env: "${ .ClusterValues.environment }"
        region: "${ .ClusterValues.region }"
      metrics:
        enabled: "${ .ClusterValues.enableMetrics }"
        # Example of using ClusterLabels alongside ClusterValues
        siteIdentifier: "${ index .ClusterLabels \"site-id\" }"
====

=== Structural Generation (Advanced)

Standard YAML blocks must be valid YAML before processing. This prevents you from using ++range++ or ++if++ to generate lists or maps. Because ++templateValues++ is a map of strings, it bypasses this restriction. {product_name} evaluates the string first and then parses the result as YAML.

The following example deploys different image tags depending on the cluster environment.

[source,yaml]
====
helm:
  templateValues:
    imageTag: ${ if eq .ClusterLabels.env "prod" }stable${ else }latest${ end }

  values:
    image:
      repository: nginx
      tag: ${ .TemplateValues.imageTag }
====

This results in

> * If the cluster label is ++env=prod++, then the gendered image tab is ++stable++.

Reviewer: "gendered image tab"? 🤔 
Do we mean the following? (same remark on the next line)
```suggestion
* If the cluster label is ++env=prod++, then the rendered image tag is ++stable++.
```

Author: Yes, fixed.
